### ISSUE TITLE
0.1.x Proposal: Restructing of data structures

### ISSUE THREAD

**Issue by frewsxcv** [georust/rust-geojson#29]
@georust/core Would be great to get opinions on this:

I'm not very satisfied with how this project as it's data laid out. Right now:

* Every single type has its own `struct`. This has one large disadvantage. For `Geometry` types like `Point`, `MultiPoint`, `LineString`, etc., they all share the same extra data: an optional `Crs` and an optional `Bbox`. Those two fields are not yet implemented, but if they were, it'd require an extra two fields on every single `struct`. This is not bad, but it's not great.
* We have two extra types that aren't really part of the spec: `Ring` and `Poly`. These are largely unnecessary and make the library less ergonomic since it requires an additional layer of abstraction for some `Geometry` types.

I thought about this tonight and ended up restructuring everything:

```rust
pub type Bbox = Vec<f64>;

pub enum Crs {
    Named {
        name: String,
    },
    Linked {
        href: String,
        type_: Option<String>,
    },
}

pub type Position = Vec<f64>;

pub type Point = Position;
pub type LineString = Vec<Position>;
pub type Polygon = Vec<Vec<Position>>;

pub enum Value {
    Point(Point),
    MultiPoint(Vec<Point>),
    LineString(LineString),
    MultiLineString(Vec<LineString>),
    Polygon(Polygon),
    MultiPolygon(Vec<Polygon>),
    GeometryCollection(Vec<Geometry>),
}

pub struct Geometry {
    pub bbox: Option<Bbox>,
    pub value: Value,
    pub crs: Option<Crs>,
}

pub struct Feature {
    pub bbox: Option<Bbox>,
    pub crs: Option<Crs>,
    pub geometry: Geometry,
    pub id: Option<String>,
    pub properties: Option<json::Object>,
}

pub struct FeatureCollection {
    pub bbox: Option<Bbox>,
    pub crs: Option<Crs>,
    pub features: Vec<Feature>,
}

pub enum GeoJson {
    Geometry(Geometry),
    Feature(Feature),
    FeatureCollection(FeatureCollection),
}
```

Big differences:

* As far as I know, the new structure covers the entire GeoJSON spec!
* `Bbox` and `Crs` are added
 * These are getting added regardless of the restructure
* `Geometry` `struct` and `Value` `enum`
 * To prevent redundancy, I put the common fields in `Geometry` and the individual data in their respective `enum` variant
* `Point`, `LineString`, and `Polygon` types
 * These sort of replace `Ring` and `Poly`. They mainly exist so we don't have to write `Vec<Vec<Vec<Positions>>>` everywhere.

Please let me know any opinions, concerns, thoughts on this. If I don't hear anything within a week, I'll move forward with this and bump to version `0.1.x`.

EDITS:

* Wrapped `Feature::properties` in `Option`
* Renamed `Geometry::_type` to `Geometry::coordinates`
* Renamed `Positions` to `Position`
* `GeometryCollection(Vec<GeometryType>)` -> `GeometryCollection(Vec<Geometry>)`
* Rename `GeometryType` to `Value`, `Geometry::type_` to `Geometry::value`
* For the sake of better ergonomics, change `Bbox` and `Position` to be `type`s instead of `struct`s
 * Before: `Value::Point(Position(vec![1.0, 2.0, -3.0]))`
 * After: `Value::Point(vec![1.0, 2.0, -3.0])`
* Add `GeoJson` enum
 * This is where `from_json` and `to_json` will go as the main entry point API for this library. Just like the spec, every valid GeoJson object is either a `Geometry`, a `Feature`, or a `FeatureCollection`.
* Add `pub` prefixes

**Comment by skj3gg** [comment]
> I thought about this tonight and ended up restructuring everything

Lmao you are funny... and amazing! I'll look tomorrow but nothing is jumping out as intrinsically bad.

**Comment by mgax** [comment]
I like it! A couple of comments:

* How about renaming `Geomentry.type_` to `Geometry.coordinates`, to clarify and bring it in line with the GeoJSON naming?
* I think `Feature.propeties` should be `Option<json::Object>`, to make the distinction between no properties and an empty properties object.

**Comment by frewsxcv** [comment]
@mgax Great ideas, thanks! I just updated the code block above with your suggestions.

**Comment by sunng87** [comment]
:+1: for `Geometry.coordinates`

This design looks good from me

**Comment by skj3gg** [comment]
Could you throw it into a branch so I could test with it, or is it not fully complete?

**Comment by frewsxcv** [comment]
I just put it in the Rust playpen: http://is.gd/w60ECw

I commented out `Feature::properties` since the `rustc_serialize::json` library+module is not available

**Comment by frewsxcv** [comment]
One minor downside of the `type_` -> `coordinates` renaming is that it's misleading if one uses `GeometryType::GeometryCollection` since the array in a `GeometryCollection` is titled `geometries`, not `coordinates`.

On a side note, if anyone has a better name for `GeometryType`, please let me know. I'm not too satisfied with it and it'd be great if we could think of something shorter since it'll get typed often. Some ideas: `GeomType`, `Type`, or `Geometry` (and rename the existing `Geometry` to something else)

```
Geometry {coordinates: GeometryType::GeometryCollection(vec![])}
```

**Comment by mgax** [comment]
How about `enum Value` and `Geometry.value`?

**Comment by frewsxcv** [comment]
> How about enum Value and Geometry.value?

I like those options more than `enum GeometryType` and `geometry.coordinates`, but I am going to leave this issue open for a few more days in case we think of anything better.

**Comment by frewsxcv** [comment]
Gave this some more thought over the weekend. Made a few changes to the codeblock in my original post in this Issue (I documented what I changed/added). If I don't hear anything for a couple days, I'll start implementing this Tuesday night in a new branch.

**Comment by frewsxcv** [comment]
Just tagged and released 0.1.0
